fix: clean up logs job was broken

shadrak98 opened the pull request:
### Changes:
- Removed priority condition from `clear_outbox` since all the emails are by default on priority 1 (except newsletter which is 0). This can be debated further if we should keep it or not.
- This also resolves #15620 
- Last changes made here in [#14441](https://github.com/frappe/frappe/pull/14441) where throwing SQL syntax error whenever `log_settings.run_log_clean_up` job was triggered.
<img width="1086" alt="Screenshot 2022-02-24 at 10 35 24 AM" src="https://user-images.githubusercontent.com/30501401/155461383-c440865c-bf6b-4e83-a393-16d0ad06b05b.png">
- I couldn't figure out what was wrong with the syntax so tweaked it a bit :D


Here is the traceback:

```bash
Traceback (most recent call last):
  File "apps/frappe/frappe/core/doctype/scheduled_job_type/scheduled_job_type.py", line 80, in execute
    frappe.get_attr(self.method)()
  File "apps/frappe/frappe/core/doctype/log_settings/log_settings.py", line 35, in run_log_clean_up
    doc.clear_logs()
  File "apps/frappe/frappe/core/doctype/log_settings/log_settings.py", line 15, in clear_logs
    self.clear_error_logs()
  File "apps/frappe/frappe/core/doctype/log_settings/log_settings.py", line 21, in clear_error_logs
    frappe.db.delete(table, filters=(
  File "apps/frappe/frappe/database/database.py", line 1032, in delete
    return self.sql(query, values, **kwargs)
  File "apps/frappe/frappe/database/database.py", line 157, in sql
    self._cursor.execute(query)
  File "env/lib/python3.9/site-packages/pymysql/cursors.py", line 148, in execute
    result = self._query(query)
  File "env/lib/python3.9/site-packages/pymysql/cursors.py", line 310, in _query
    conn.query(q)
  File "env/lib/python3.9/site-packages/pymysql/connections.py", line 548, in query
    self._affected_rows = self._read_query_result(unbuffered=unbuffered)
  File "env/lib/python3.9/site-packages/pymysql/connections.py", line 775, in _read_query_result
    result.read()
  File "env/lib/python3.9/site-packages/pymysql/connections.py", line 1156, in read
    first_packet = self.connection._read_packet()
  File "env/lib/python3.9/site-packages/pymysql/connections.py", line 725, in _read_packet
    packet.raise_for_error()
  File "env/lib/python3.9/site-packages/pymysql/protocol.py", line 221, in raise_for_error
    err.raise_mysql_exception(self._data)
  File "env/lib/python3.9/site-packages/pymysql/err.py", line 143, in raise_mysql_exception
    raise errorclass(errno, errval)
pymysql.err.ProgrammingError: (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near ')' at line 1")
```

shadrak98 committed:
fix: clean up logs job was broken

gavindsouza commented on frappe/email/queue.py:
```diff
@@ -170,7 +170,7 @@ def clear_outbox(days=None):
 
 	email_queues = frappe.db.sql_list("""SELECT `name` FROM `tabEmail Queue`
-		WHERE `priority`=0 AND `modified` < (NOW() - INTERVAL '{0}' DAY)""".format(days))
+		WHERE `modified` < (NOW() - INTERVAL '{0}' DAY)""".format(days))
```
Use qb here too?

gavindsouza commented on frappe/core/doctype/activity_log/activity_log.py:
```diff
@@ -48,6 +48,7 @@ def clear_activity_logs(days=None):
+	duration = (Now() - Interval(days=days))
 	frappe.db.delete(doctype, filters=(
-		doctype.creation < PseudoColumn(f"({Now() - Interval(days=days)})")
+		doctype.creation < duration
```
```suggestion
		doctype.creation < (Now() - Interval(days=days))
```

We don't need to assign a separate variable for this. ^ has lesser cognitive load

gavindsouza commented on frappe/core/doctype/log_settings/log_settings.py:
```diff
@@ -18,8 +18,9 @@ def clear_logs(self):
+		duration = (Now() - Interval(days=self.clear_error_log_after))
 		frappe.db.delete(table, filters=(
-			table.creation < PseudoColumn(f"({Now() - Interval(days=self.clear_error_log_after)})")
+			table.creation < duration
```
```suggestion
			table.creation < (Now() - Interval(days=self.clear_error_log_after))
```

gavindsouza commented on frappe/core/doctype/log_settings/log_settings.py:
```diff
@@ -18,8 +18,9 @@ def clear_logs(self):
 
 	def clear_error_logs(self):
 		table = DocType("Error Log")
+		duration = (Now() - Interval(days=self.clear_error_log_after))
```
```suggestion
```

gavindsouza commented on frappe/core/doctype/activity_log/activity_log.py:
```diff
@@ -48,6 +48,7 @@ def clear_activity_logs(days=None):
 	if not days:
 		days = 90
 	doctype = DocType("Activity Log")
+	duration = (Now() - Interval(days=days))
```
```suggestion
```

gavindsouza commented:
Also, @shadrak98 can you add corresponding tests to validate the changes?

shadrak98 committed:
fix: removed redundant pieces & rewrote the query with qb

shadrak98 committed:
fix: added test cases for log settings

shadrak98 committed:
chore: fix linter issues

shadrak98 commented:
> Also, @shadrak98 can you add corresponding tests to validate the changes?

Added test cases.

gavindsouza commented:
Failing MariaDB flaky tests will be addressed by #16153

gavindsouza commented:
@Aradhya-Tripathi Can you check the converted queries since this was how you had converted them earlier but moved to use PseudoColumn (in https://github.com/frappe/frappe/pull/14441)?

Aradhya-Tripathi commented on frappe/core/doctype/activity_log/activity_log.py:
```diff
@@ -49,5 +49,5 @@ def clear_activity_logs(days=None):
 		days = 90
 	doctype = DocType("Activity Log")
 	frappe.db.delete(doctype, filters=(
-		doctype.creation < PseudoColumn(f"({Now() - Interval(days=days)})")
+		doctype.creation < (Now() - Interval(days=days))
 	))
```
This changes the original query

Original query ->
`delete from "tabActivity Log" where \t\tcreation< (NOW() - INTERVAL '10' DAY)`

Query obtained using pseudocolumn ->
`DELETE FROM "tabActivity Log" WHERE "creation"<(NOW()-INTERVAL '10 DAY')`

Query when pseudocolumn is removed ->
`DELETE FROM "tabActivity Log" WHERE "creation"<NOW()-INTERVAL '10 DAY'`

shadrak98 replied:
@Aradhya-Tripathi can you tell me how to verify the raw SQL query generated by qb?

Aradhya-Tripathi replied:
You can pass ```run=False``` as a kwarg to all queries going through ```def sql```  to just return the query generated without actually executing it.

shadrak98 replied:
@Aradhya-Tripathi Yeah, you were right. Thanks for pointing this out. Can you check why the system was throwing syntax errors then?
Also, my query seems to be working fine (as intended). can you verify that? 😅

Aradhya-Tripathi replied:
Hey so your query isn't wrong ie it won't give a syntax error it just differs in logic diff -> `creation<(Now() - Interval(days = days))`

gavindsouza replied:
@Aradhya-Tripathi The converted query is correct for MariaDB/Postgres. It doesn't differ in logic.

Aradhya-Tripathi commented on frappe/core/doctype/log_settings/log_settings.py:
```diff
@@ -19,7 +18,7 @@ def clear_logs(self):
 		table = DocType("Error Log")
 		frappe.db.delete(table, filters=(
-			table.creation < PseudoColumn(f"({Now() - Interval(days=self.clear_error_log_after)})")
+			table.creation < (Now() - Interval(days=self.clear_error_log_after))
 		))
```
Same with this

surajshetty3416 committed:
Merge branch 'develop' into log-settings

surajshetty3416 commented:
@shadrak98 @gavindsouza @Aradhya-Tripathi  any update on this?

gavindsouza commented on frappe/core/doctype/log_settings/test_log_settings.py:
```diff
@@ -1,8 +1,128 @@
+	def setUpClass(cls):
+		fieldnames = ['clear_error_log_after', 'clear_activity_log_after', 'clear_email_queue_after']
+		for fieldname in fieldnames:
+			frappe.set_value("Log Settings", None, fieldname, 1)
```
This can be done in a single query by passing a `dict` instead.

gavindsouza commented on frappe/core/doctype/log_settings/test_log_settings.py:
```diff
@@ -1,8 +1,128 @@
+
+	@classmethod
+	def tearDownClass(cls):
+		if frappe.db.exists({"doctype": "Activity Log", "subject": "Test subject"}):
```
I think we can get rid of this whole method by creating a `SAVEPOINT` in the setUpClass and rolling back to it here.

ankush replied:
just inherit `FrappeTestCase` 😬 https://github.com/frappe/frappe/pull/16162

gavindsouza replied:
After it starts using savepoints instead of committing on each new case 😜 @ankush

gavindsouza commented on frappe/core/doctype/log_settings/test_log_settings.py:
```diff
@@ -1,8 +1,128 @@
+			for log in error_logs:
+				frappe.db.delete("Error Log", log)
+
+	def test_create_activity_logs(self):
```
1. `test_create_activity_logs`, `test_create_error_logs` & `test_create_email_queue` don't exactly test for those document creations accurately.
2. `doc.db_set` is much cleaner than using a set_value btw
3. These tests don't belong in `TestLogSettings`. I don't think these tests contribute to anything either. They're more like setUp methods for `test_delete_logs`

gavindsouza commented on frappe/core/doctype/log_settings/test_log_settings.py:
```diff
@@ -1,8 +1,128 @@
+
+		self.assertEqual(len(email_queues), 2)
+
+	def test_delete_logs(self):
```
`test_delete_logs` isn't an atomic unittest. It depends on the previous 3 tests to do anything. 

Also, this test has indicators of a flaky one as we don't empty the other tables before creating and testing for this.

gavindsouza committed:
test(log_settings): test_delete_logs

* Refactored TestCase to test "correctly"
* Got rid of earlier flaky logic
* Dropped unwanted tests

gavindsouza committed:
fix(activity_log): Remove unused import from namespace

gavindsouza committed:
fix(set_single_value): Make value parameter optional

Similar to set_value for accepting multiple columns ot be updated for
the same Table through a Dict as the second positional arg

Misc: Added type hints

gavindsouza committed:
refactor(minor): clear_outbox

* Use pluck API instead of building dict and then accesing keys
* Styled query
* Added type hints

gavindsouza committed:
Merge branch 'develop' into log-settings

muchai commented:
Will this be backported to version - 13 at some point? I see the label of do not backport.

ankush commented:
@muchai this doesn't affect v13.

muchai commented:
@ankush It does. Had reported it here: https://github.com/frappe/frappe/issues/15620

ankush commented:
@shadrak98 can you confirm if this affects v13 too? 👀 Maybe parts of this PR need to be ported.

muchai commented:
Please see if you can backport parts fixing log settings deletion.

ankush commented:
@muchai can you please elaborate which logs aren't getting deleted for you right now?

muchai commented:
@ankush Email Queue, or at least that's the biggest consumer of space.

ankush commented:
@muchai but is old email queue present on your site? (older than configured # of days)

muchai commented:
@ankush affirmative.
Log Settings
![image](https://user-images.githubusercontent.com/1822868/169973080-1ad8003f-84a0-4c9e-9010-b9265d5c8e3b.png)

Email Queue, records from eleven months ago
![image](https://user-images.githubusercontent.com/1822868/169973225-60dc5cce-2006-4d22-b6cc-7fbac34447b2.png)

ankush commented:
@muchai I have a hunch that the job that's supposed to delete it might be failing on your site because of huge old data?

Can you check "scheduled job log" for job type "log_settings.run_log_clean_up` ?

ankush commented:
@muchai continue on this issue: https://github.com/frappe/frappe/issues/16971

shadrak98 commented:
@ankush 
@muchai as suggested this PR will be helpful if backported to v13 since not all email queues are deleted, only once which has priority set as 0 (i.e newsletters) rest are ignored. 
https://github.com/frappe/frappe/blob/version-13/frappe/email/queue.py#L595-L596

Also, if backported it won't help much since this only deletes queues for certain days back for e.g 31 days old email queues, we might need a patch to clear all old email queues as well?
<img width="627" alt="Screenshot 2022-05-24 at 1 24 58 PM" src="https://user-images.githubusercontent.com/30501401/169979329-8934781d-fcbc-4fdb-8648-169a9ddebc6a.png">
